packages/parser/lib/driver_.re

```diff
@@ -0,0 +1,56 @@
+module Lexer = Css_lexer;
+module Parser = Css_parser;
+module Location = Ppxlib.Location;
+
+type parser('token, 'ast) = MenhirLib.Convert.traditional('token, 'ast);
+exception LexingError((Lexing.position, string));
+
+let menhir = MenhirLib.Convert.Simplified.traditional2revised;
+
+let parse = (skip_whitespaces, lexbuf, parser) => {
+  Lexer.skip_whitespace.contents = skip_whitespaces;
+  let last_token = ref((Parser.EOF, Lexing.dummy_pos, Lexing.dummy_pos));
+
+  let next_token = () => {
+    last_token := Lexer.get_next_tokens_with_location(lexbuf);
+    last_token^;
+  };
+
+  try(Ok(menhir(parser, next_token))) {
+  | LexingError((pos, msg)) =>
+    let loc = Lex_buffer.make_loc(pos, pos);
+    Error((loc, msg));
+  | _ =>
+    let (token, start_pos, end_pos) = last_token^;
+    let loc = Lex_buffer.make_loc(start_pos, end_pos);
+    let msg =
+      Printf.sprintf(
+        "Parse error while reading token '%s'",
+        Tokens.token_to_string(token),
+      );
+    Error((loc, msg));
+  };
+};
+
+// TODO: Don't pass ~container_lnum and pos around, handle location all in here.
```

Comment:
As mentioned, container_lnum was used to print to errors on the right locations.

We don't have tests around this, but would be nice to eventually fix this.

Can you add the reverse todo?

Reply:
Yes.
Generally, the PR need some bit of polishing